Test(precision): add theoretical accuracy test code

grysgreat opened the pull request:
in cubic_spline interpolation function
which covers cubic_spline interpolation functions
in class cubic_spline and math_zone_add1.
Test cases containing four boundary cases in uniform 
and non-uniform sampling --Issue #2837

grysgreat committed:
in cubic_spline interpolation function
which covers cubic_spline interpolation functions
in class cubic_spline and math_zone_add1.
Test cases containing four boundary cases in uniform and non-uniform sampling
--Issue #2837

grysgreat committed:
in cubic_spline interpolation function
which covers cubic_spline interpolation functions
in class cubic_spline and math_zone_add1.
Test cases containing four boundary cases in uniform and non-uniform sampling
--Issue #2837

grysgreat committed:
Test(precision): fix a bug in last commit
(due to some code change in early pr in cubic_spline class)
--issue#2894

grysgreat committed:
Test(precision): fix a bug in last commit
(due to some code change in early pr in cubic_spline class)
--issue#2894

grysgreat committed:
Test(precision): fix a test bug in exp testcase.
Sample the function exp(x) uniformly under condition not_a_knot can not pass.
The error is about 1e-6(which is very small)

grysgreat committed:
Test(precision): fix a test bug in exp testcase.
Sample the function exp(x) uniformly under condition not_a_knot can not pass.
The error is about 1e-6(which is very small)

jinzx10 commented on source/module_base/test/cubic_spline_test.cpp:
1. why the test on static member function is removed?
2. why the tolerance is changed from 1e-15 to 1e-8?

There are a few possible ways to test the code, e.g. (a) compare results from the current implementation with those from other implementation, and (b) compare results with theoretical error bound. The previous code does (a), compares the result with scipy.integrate.CubicSpline, and set the tolerance to 1e-15. The purpose is to check that it does exactly the same job as scipy does, and this tolerance should be understood as an "floating-point tolerance".

jinzx10 commented on source/module_base/test/cubic_spline_test.cpp:
the design of "count_err" is not good practice; I would suggest replacing "int func" with functor "std::function<double(double)>", creating lambdas local to each test, and passing that lambda to count_err.

jinzx10 replied:
1. why the test on static member function is removed?
2. why the tolerance is changed from 1e-15 to 1e-8?

There are a few possible ways to test the code, e.g. (a) compare results from the current implementation with those from other implementation, and (b) compare results with theoretical error bound. The previous code does (a), compares the result with scipy.integrate.CubicSpline, and set the tolerance to 1e-15. The purpose is to check that it does exactly the same job as scipy does, and this tolerance should be understood as an "floating-point tolerance".

grysgreat committed:
Merge branch 'deepmodeling:develop' into precision_testing

grysgreat committed:
Test(precision):fix some mistake mentioned earlier.
For question 1, this is because I tried to test the single-precision code earlier, so I adjusted the error threshold to 1e-7.
For question 2, this is because there was a new pr commit for the test when I changed the code, and I didn't notice that a new function had been added. Both issues have been fixed.
--issue #2894

grysgreat committed:
Test(precision):fix some mistake mentioned earlier.
For question 1, this is because I tried to test the single-precision code earlier, so I adjusted the error threshold to 1e-7.
For question 2, this is because there was a new pr commit for the test when I changed the code, and I didn't notice that a new function had been added. Both issues have been fixed.
--issue #2894

grysgreat committed:
Merge branch 'precision_testing' of https://gitee.com/guhongyaoxing/abacus-grys into precision_testing

grysgreat commented:
fix 2 bugs in last commented mentioned @jinzx10 (I don't seem to have found your comment on github ... )

jinzx10 reviewed (changes requested):
Why the test "expx_Boundary2" is added but commented out?

I think the tests for the old interface (SplineD2) do not apply the natural boundary condition (f''(0)=f''(n)=0) correctly. The old interface does have a weird behavior that you should use with care. If I remember correctly, it is designed such that natural boundary condition is invoked when the 4th and 5th arguments are larger than 99999 (you may read the code to confirm that).

grysgreat committed:
Test(precision): use lamba expression function
to calculate parameter passing for theoretical error. (adopted comment by jinzx10)
--Issue #2837

grysgreat committed:
Test(precision): use lamba expression function
to calculate parameter passing for theoretical error. (adopted comment by jinzx10)
--Issue #2837

grysgreat committed:
Test(precision): fix a bug in previous  commit.
--Issue #2837

grysgreat committed:
Merge branch 'precision_testing' of https://gitee.com/guhongyaoxing/abacus-grys into precision_testing

grysgreat commented:
For your first question:
According to the theory I have found, the theoretical error estimate is only suitable for cubic spline interpolation functions of the first or second kind of boundary conditions, i.e. when the exact value of the boundary derivative is known. For natural boundary conditions, such an estimated derivative may lead to large errors.
Here, for the natural boundary condition interpolation of exp function, due to the large boundary derivative, the theoretical error test result does not pass, so I According to the theory I have found, the theoretical error estimate is only suitable for cubic spline interpolation functions of the first or second kind of boundary conditions, i.e. when the exact value of the boundary derivative is known. For natural boundary conditions, such an estimated derivative may lead to large errors.
Here, for the natural boundary condition interpolation of exp function, due to the large boundary derivative, the theoretical error test result does not pass, so I comment it.

grysgreat commented:
> Why the test "expx_Boundary2" is added but commented out?
> 
> I think the tests for the old interface (SplineD2) do not apply the natural boundary condition (f''(0)=f''(n)=0) correctly. The old interface does have a weird behavior that you should use with care. If I remember correctly, it is designed such that natural boundary condition is invoked when the 4th and 5th arguments are larger than 99999 (you may read the code to confirm that).

The implementation of natural conditions about this function is just as you said.
According to the theory I have found, the theoretical error estimate is only suitable for cubic spline interpolation functions of the first or second kind of boundary conditions, when the exact value of the boundary derivative is known, for the natural boundary condition interpolation of exp function, due to the large boundary derivative, the theoretical error test result does not pass, so I comment it.

grysgreat commented:
> the design of "count_err" is not good practice; I would suggest replacing "int func" with functor "std::function<double(double)>", creating lambdas local to each test, and passing that lambda to count_err.

Indeed, it is more elegant to use lambda expressions to pass function arguments, and this approach has been adopted in the new PR.

jinzx10 commented:
> > Why the test "expx_Boundary2" is added but commented out?
> > I think the tests for the old interface (SplineD2) do not apply the natural boundary condition (f''(0)=f''(n)=0) correctly. The old interface does have a weird behavior that you should use with care. If I remember correctly, it is designed such that natural boundary condition is invoked when the 4th and 5th arguments are larger than 99999 (you may read the code to confirm that).
> 
> The implementation of natural conditions about this function is just as you said. According to the theory I have found, the theoretical error estimate is only suitable for cubic spline interpolation functions of the first or second kind of boundary conditions, when the exact value of the boundary derivative is known, for the natural boundary condition interpolation of exp function, due to the large boundary derivative, the theoretical error test result does not pass, so I comment it.

Could you provide the reference for the error bound? I found a paper that seems to be relevant to your test & code

Charles A Hall, W.Weston Meyer, Optimal error bounds for cubic spline interpolation
https://www.sciencedirect.com/science/article/pii/002190457690040X

but the test code does not exactly perform tests as suggested by the paper. 

1. I find Eq. 8 in the above paper agrees with your function "count_err" when the 4th argument is 0, but the constant prefactor is no longer 5/384 for non-zero derivative. I do not see count_err used with non-zero 4th argument, but I still would like to know if your code is indeed backed by some literature or just some mistake/typo. 
2. The above paper discussed the error bound in terms of type-I / type-II interpolant, i.e., interpolants whose boundary first or second derivative agree with those of the underlying function. If that's the case, not-a-knot boundary condition should not be tested under this error bound since the paper say nothing about that; pass or failed could just be some coincidence. I notice that NotAKnot_expx_uniform is commented out but some other not-a-knot boundary condition is tested. I think if there's no theoretical error bound for that, you may skip its test at all (instead of using some accidentally passed tests).
3. If I understand correctly, the functor to passed to count_err is supposed to be the fourth-derivative of the underlying function. But why a constant function auto f = [](double x) -> double { return 1; } is passed to count_err in the case of f(x)=2*x ? 
4. I think it might be better if your code could have some comments to explain what's going on. If the code is doing a bunch of similar things, a good practice is to have a big comment block at the beginning to introduce them all at once.

grysgreat committed:
Merge branch 'deepmodeling:develop' into precision_testing

grysgreat committed:
Test(precision): fix some typo error, and add notes.

grysgreat committed:
Merge branch 'precision_testing' of https://gitee.com/guhongyaoxing/abacus-grys into precision_testing

grysgreat commented:
>but the test code does not exactly perform tests as suggested by the paper.

>I find Eq. 8 in the above paper agrees with your function "count_err" when the 4th argument is 0, but the constant prefactor is no longer 5/384 for non-zero derivative. I do not see count_err used with non-zero 4th argument, but I still would like to know if your code is indeed backed by some literature or just some mistake/typo.


Here are the materials and papers I refer to:
https://math.ecnu.edu.cn/~sfzhu/course/NumerAnal/Interpolation5.pdf
https://www.math.pku.edu.cn/teachers/litj/notes/numer_anal/CubicSplineErrorBounds.pdf

1. According to the paper, the error analysis is the same as what you see. On this issue, in fact, the derivative order option was designed according to the formula in the paper in the early design of this function. Later, it was found that the spline function of abacus would only return the original value, so this parameter was not used, so the constant term had only one value 5/384.

2. Tests that are not a knot condition can indeed be omitted (the test sample for exp function was annotated because the test failed), and these test samples can be annotated at the subsequent commit.

3. This was indeed a clerical error and has been fixed.

4. Comments are added.
